chronograf.go

```diff
@@ -625,3 +625,66 @@ type LayoutStore interface {
 	// Update the dashboard in the store.
 	Update(context.Context, Layout) error
 }
+
+// SourceAndKapacitor allows a --new-source to be parsed into a source and a kapacitor
+type SourceAndKapacitor struct {
+	Source    Source `json:"influxdb"`
+	Kapacitor Server `json:"kapacitor"`
+}
+
+// NewSources adds sources and respective kapacitors to BoltDb idempotently by name
+func NewSources(ctx context.Context, sourcesStore SourcesStore, serversStore ServersStore, newSources string, logger Logger) error {
+	// Once a bolt connection is created, try to add any new sources with respective kapacitor servers
+	if newSources != "" {
+		var srcsKaps []SourceAndKapacitor
+		// On JSON unmarshal error, continue server process without new source and write error to log
+		if err := json.Unmarshal([]byte(newSources), &srcsKaps); err != nil {
+			logger.
+				WithField("component", "server").
+				WithField("NewSource", "invalid").
+				Error(err)
+		} else {
+			srcs, err := sourcesStore.All(ctx)
+			if err != nil {
+				return err
+			}
+			kaps, err := serversStore.All(ctx)
+			if err != nil {
+				return err
+			}
+
+			for _, srcKap := range srcsKaps {
+				isNewSource := true
+				for _, src := range srcs {
+					if src.Name == srcKap.Source.Name {
+						isNewSource = false
+						break
+					}
+				}
+				if isNewSource == true {
+					src, err := sourcesStore.Add(ctx, srcKap.Source)
+					if err != nil {
+						return err
+					}
+
+					isNewKapacitor := true
```

Comment:
I don't think it's necessary to check whether the kapacitor exists. A kapacitor in BoltDB is really just a configuration for a kapacitor instance, and in theory you could have multiple InfluxDB sources that are all configured to work with the same kapacitor instance. Kapacitor can be configured to work with multiple InfluxDB servers, so this is actually a valid use case.

Kapacitor configs belong to sources, so a newly created source will not have any. It's fine in this case to just create a new kapacitor for that source, using the provided configuration.